### PR TITLE
feat(core)!: Change data processing for multi-input-nodes

### PR DESCRIPTION
**Pull request by janober**
This PR contains two changes that make it possible to configure how nodes with multiple inputs behave. It is now possible to control if incoming nodes get forced to execute (currently default), and it is possible to define which inputs are required.

Important to mention that the setting `forceInputNodeExecution` always force executes all nodes from all inputs rather than just the ones defined in `requiredInputs`. (Background: this specific behavior is currently not required and may never be, so no need to spend time on it and unnecessarily complicate the code)

## Example of how new versions of the Merge-Node will behave now

Current data processing behavior:
Screenshot from 2022-10-03 22-34-26

New data processing behavior:
Screenshot from 2022-10-03 22-35-42

Community topics:
https://community.n8n.io/t/merge-data-node-getting-triggered-even-without-data-input-from-both-source/22174
https://community.n8n.io/t/getting-error-in-if-node/22179
https://community.n8n.io/t/if-node-is-true-and-false-in-one-execution/3169/5
https://community.n8n.io/t/if-node-continues-to-both-branches-even-though-it-is-false-and-data-is-on-the-false-side/21908
https://community.n8n.io/t/merge-wait-seems-broken/18725
https://community.n8n.io/t/unexpected-behavior-from-unvisited-nodes-in-a-switch-and-merge-workflow/10278
https://community.n8n.io/t/merge-node-suggestion-add-option-to-disable-triggering-of-previous-nodes/9931/13
https://community.n8n.io/t/workflow-nodes-execution-order/13546/6
https://community.n8n.io/t/bug-merge-node-makes-all-preceding-nodes-execute/25939
https://community.n8n.io/t/merge-node-merging-multiple-outputs/19580/9
https://community.n8n.io/t/merge-behavior-not-expected/26226
https://community.n8n.io/t/merge-node-choosebranch-behaviour/27579/2

GitHub issues:
https://github.com/n8n-io/n8n/issues/6596
https://github.com/n8n-io/n8n/issues/3949

### CODE REVIEW

File: packages/cli/src/databases/migrations/sqlite/index.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@ import { IntroducePinData1654089251344 } from './1654089251344-IntroducePinData'
 import { AddNodeIds1658930531669 } from './1658930531669-AddNodeIds';
 import { AddJsonKeyPinData1659888469333 } from './1659888469333-AddJsonKeyPinData';
 import { CreateCredentialsUserRole1660062385367 } from './1660062385367-CreateCredentialsUserRole';
+import { AddWorkflowVersion1664567980717 } from './1664567980717-AddWorkflowVersion';
```
Comment (flipswitchingmonkey) [inline packages/cli/src/databases/migrations/sqlite/index.ts:20]: Missing Postgres and MySQL migrations

File: packages/core/src/WorkflowExecute.ts [hunk 2]
```diff
@@ -1195,6 +1209,113 @@ export class WorkflowExecute {
...
+							if (found) {
+								// Execute node later as one of its dependencies is still outstanding
+								continue;
+							}
+
+							const runIndexes = Object.keys(
+								this.runExecutionData.executionData!.waitingExecution[nodeName],
+							).sort();
+
+							const firstRunIndex = parseInt(runIndexes[0]);
```
Comment (flipswitchingmonkey) [inline packages/core/src/WorkflowExecute.ts:1270]: Are we certain that runIndexes.length is always > 0 here?

**Review by flipswitchingmonkey — CHANGES_REQUESTED** [review]

File: packages/cli/commands/worker.ts [hunk 3]
```diff
@@ -175,6 +175,7 @@ export class Worker extends Command {
 
 		const workflow = new Workflow({
 			id: currentExecutionDb.workflowData.id as string,
+			version: currentExecutionDb.workflowData.version as number,
```
[the remark this answers is not in the document]
Comment (janober) [reply packages/cli/commands/worker.ts]: Workflow-Version got removed so comment is outdated.

File: packages/cli/src/databases/migrations/sqlite/index.ts (hunk 1 above)
Comment (janober) [reply packages/cli/src/databases/migrations/sqlite/index.ts]: Workflow-Version got removed so the comment is outdated.

File: packages/cli/src/databases/migrations/sqlite/1664567980717-AddWorkflowVersion.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+		await queryRunner.query(
+			`ALTER TABLE \`${tablePrefix}workflow_entity\` ADD COLUMN "version" integer`,
+		);
+
+		logMigrationEnd(this.name);
+	}
+
+	async down(queryRunner: QueryRunner): Promise<void> {
+		// TODO: Add the proper down migration
```
[the remark this answers is not in the document]
Comment (janober) [reply packages/cli/src/databases/migrations/sqlite/1664567980717-AddWorkflowVersion.ts]: Workflow-Version got removed so the comment is outdated.

File: packages/core/src/WorkflowExecute.ts (hunk 2 above)
Comment (janober) [reply packages/core/src/WorkflowExecute.ts:1304]: Yes to 100%. After all, do is `runIndexes` generated from `this.runExecutionData.executionData.waitingExecution` and data from it is what we iterate over.

**Comment by edumaciel10** [comment]
Great PR, but I have one doubt, how I can test your pr locally?

**Comment by janober** [comment]
@edumaciel10 you can test it with the following docker command:
```
docker run -it --rm --name n8n -p 5678:5678 -v ~/.n8n:/home/node/.n8n n8nio/n8n:PR-4238-workflow-version-2
```

**Comment by janober** [comment]
@renattomachado  It will be released with n8n v1.0 end of June.

File: packages/nodes-base/nodes/Merge/v2/MergeV2.node.ts [hunk 5]
```diff
@@ -473,6 +485,37 @@ export class MergeV2 implements INodeType {
...
+					'Input 1',
+				);
+				if (!input1) return [returnData];
+
+				input2 = checkInput(
+					input2,
+					matchFields.map((pair) => pair.field2),
+					options.disableDotNotation || false,
+					'Input 2',
+				);
```
[the remark this answers is not in the document]
Comment (janober) [reply packages/nodes-base/nodes/Merge/v2/MergeV2.node.ts]: Another good find! I must have copied that over without thinking when I merged and reintroduced the changes required for that new functionality. Got removed.

**Review by maspio — APPROVED** [review]

**Review by netroy — APPROVED** [review]

**Review by netroy — CHANGES_REQUESTED** [review]
Not sure why, but these changes are [causing issues in the QuestDB node in one of the test workflows](https://github.com/n8n-io/n8n/actions/runs/5205098008/jobs/9390323840#step:11:9488).

image

**Comment by netroy** [comment]
Still seeing errors in workflow tests. 
image

**Review by netroy — APPROVED** [review]

**Comment by netroy** [comment]
Waiting on [workflow tests](https://github.com/n8n-io/n8n/actions/runs/5321117525/jobs/9635801802) and [e2e tests](https://github.com/n8n-io/n8n/actions/runs/5321115192?pr=4238) to pass. We can then merge.

**Review by netroy — APPROVED** [review]

**Review by netroy — APPROVED** [review]

**Review by netroy — APPROVED** [review]
The test-workflows related issues seem to be from `EXECUTIONS_PROCESS` being set to `main` now. Everything seems to work as expected in `own` mode. 
Merging this, and will try to fix the workflow issues in the release branch.

**Comment by janober** [comment]
Got released with [`n8n@1.0.5`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.0.5)

**Pull request merged**